ex2/sawada/main.py

```diff
@@ -0,0 +1,249 @@
+    Returns:
+        ndarray: filter
+    """
+    
+    regularized_cutoff_frequency = 2 * np.pi * cutoff_frequency / sr
+    t = np.arange(nperseg)
+    
+    # create filter
+    filter = 2 * regularized_cutoff_frequency \
+        * np.sinc(regularized_cutoff_frequency / np.pi * (t - nperseg // 2)) \
+        * np.hamming(nperseg)
+    return filter
+
+
+def create_lpf(cutoff_frequency, fft_size=512, sr=16000):
+    """low-pass filter
+
+    Args:
+        cutoff_frequency (int): cutoff frequency
+        fft_size (int, optional): samples in stft segment. Defaults to 512.
+        sr (int, optional): sampling rate. Defaults to 16000.
+
+    Returns:
+        ndarray: low-pass filter
+    """
+    return _create_base_filter(cutoff_frequency, fft_size, sr)
+
+
+def create_hpf(cutoff_frequency, fft_size=512, sr=16000):
+    """high-pass filter
+
+    Args:
+        cutoff_frequency (int): cutoff frequency
+        fft_size (int, optional): samples in stft segment. Defaults to 512.
+        sr (int, optional): sampling rate. Defaults to 16000.
+
+    Returns:
+        ndarray: high-pass filter
+    """
+    return _create_base_filter(sr / (2 * np.pi)) \
+        - _create_base_filter(cutoff_frequency, fft_size, sr)
+
+
+def apply_filter(audio, filter):
+    """apply filter to audio
+
+    Args:
+        audio (ndarray): audio array
+        filter (ndarray): filter
+
+    Returns:
+        ndarray: filtered audio
+    """
+    # define variables often used
+    filter_length = len(filter)
+    filter_fliped = filter[::-1]
+    
+    # initialize return value
+    result = np.zeros(len(audio) + filter_length - 1)
+    
+    audio_extended = np.concatenate([
+        np.zeros(filter_length - 1),
+        audio,
+        np.zeros(filter_length - 1)])
+
+    # convolve
+    for i in range(len(result)):
+        result[i] = np.sum(filter_fliped * audio_extended[i: i + filter_length])
+    return result
+
+
+def stft(audio, fs, nperseg=512, noverlap=256):
+    """Short-Time Fourier Transform
+
+    Args:
+        audio (array_like): input audio array
+        fs (float): sampling rate of input audio
+        nperseg (int, optional): samples in stft segment. Defaults to 512.
+        noverlap (int, optional): samples of stft overlap. Defaults to 256.
+
+    Returns:
+        ndarray: stft of audio
+        ndarray: time axis
+        ndarray: frequency axis
+    """
+
+    # zero padding at end of audio
+    padding_length = nperseg - len(audio) % (nperseg - noverlap)
+    audio = np.concatenate([audio, np.zeros(padding_length)])
+
+    # create window (hanning)
+    window = np.hanning(nperseg)
+
+    # stft
+    steps = (len(audio) - noverlap) // (nperseg - noverlap)
+    Zxx = np.empty((0, nperseg))
+    for i in range(steps):
+        # apply window
+        sample = window * audio[
+            (nperseg - noverlap) * i: (nperseg - noverlap) * i + nperseg
+        ]
+        # fft
+        Zxx = np.append(Zxx, [np.fft.fft(sample)], axis=0)
```

Comment:
append は使わずに、(あらかじめ全体の大きさを計算して 0初期化)→(部分配列に代入)という形式に直してください。

以下は for なしバージョンです。（これにはしなくてもいいですがぜひ一読を）
```suggestion
    from numpy.lib.stride_tricks import as_strided
    # for 文なしでフレーミング
    framed_audio = as_strided(audio,
      shape=(nperseg,steps), 
      strides=(audio.strides[0],audio.strides[0]*(nperseg-noverlap)))
    # ブロードキャストできるように窓に軸を追加して窓かけ
    windowed_audio = framed_audio*window[:,np.newaxis]
    # fast fourier transform
    Zxx = np.fft.fft(windowed_audio, axis=0)
```

Reply:
これ, 前回のSTFTの発表時に指摘頂いたところを直し忘れたままコピペしてもってきていましたね…

上のコメントも合わせて, 「for文は行列演算に置き換えれる」, アタマに叩き込んでおきます.

Reply:
あ、せっかくなので前回の .py ファイルから STFT 関数をインポートしてくる感じで書くのもいいと思います。修正は前回の方に行う感じで